src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingSupport.java

```diff
@@ -480,8 +480,14 @@ public static boolean isAssignableTo(ClassNode type, ClassNode toBeAssignedTo) {
         return false;
     }
 
-    static boolean isVargs(final Parameter[] parameters) {
+    static boolean isVargs(final Parameter[] parameters, ClassNode[] argumentTypes) {
         if (parameters == null || parameters.length == 0) return false;
+
+        // GROOVY-10056: Inferred parameter type of lambda expression for multi-dimensions array is not correct
+        if (null != argumentTypes && argumentTypes.length == parameters.length && argumentTypes[argumentTypes.length - 1].isArray()) {
+            return false;
+        }
+
```

Comment:
Instead of changing the definition of what is a variadic method, can you change the lastArgMatchesVarg result only?  A method is variadic regardless of the arguments passed.

Reply:
I tried to revert all changes and just to refine the check as you suggested:

**Original**
https://github.com/apache/groovy/blob/master/src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java#L5286

**Refined**
```
boolean isVargs = -1 != lastArgMatchesVarg(parameters, expressions.stream().map(e -> getDeclaredOrInferredType(e)).toArray(ClassNode[]::new));
```

but the build will be broken, then we can not debug... 
`lastArgMatchesVarg` is much more complicated than `isVargs`, it is hard to replace `isVargs`.

Reply:
In order to keep the method `isVargs` as it is, I even tried the following approach, but the build will still be broken...

**Original**
https://github.com/apache/groovy/blob/master/src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java#L5286

**Refined**
```java
        boolean isVargs = isVargs(parameters);
        if (isVargs) {
            ClassNode[] argumentTypes = expressions.stream().map(e -> getDeclaredOrInferredType(e)).toArray(ClassNode[]::new);
            if (null != argumentTypes && argumentTypes.length == parameters.length && argumentTypes[argumentTypes.length - 1].isArray()) {
                isVargs = false;
            }
        }
```

Reply:
I could take a look at this case when I get through GROOVY-10053 and GROOVY-10054.  I'm trying to look over all the generics fallout from 4.0a3 as a whole.

Reply:
OK. Thanks.
I'll close the PR without merging as some further tweak could be done.